Add custom templating support

pseudomorph opened the pull request:
This PR is a fork of the changes from #2481 , and addresses linting errors.

https://github.com/runatlantis/atlantis/pull/2541#issue-1387584043

pseudomorph commented:
@jamengual - Thanks for your response. I'll try to work on this later today.

I forked from @ribejara-te 's PR since this is a feature I'd really like to see. I see the pre-release was already cut. Is there any chance of getting it added in to the upcoming release if I can address the documentation, or would it need to wait until a future release?

nitrocode commented:
@pseudomorph we do multiple pre-releases. We will review this before the next pre-release.

jamengual commented:
> @jamengual - Thanks for your response. I'll try to work on this later today.
> 
> I forked from @ribejara-te 's PR since this is a feature I'd really like to see. I see the pre-release was already cut. Is there any chance of getting it added in to the upcoming release if I can address the documentation, or would it need to wait until a future release?

we will do another prerelease in like 1-2 weeks so we can add this in.

nitrocode commented:
@pseudomorph could you add a test for a custom template as well?

pseudomorph commented:
I've added a flag/config parameter for setting where the custom template directory is. Added a single test to validate that a custom template gets rendered.

Will try to get to the documentation in the within the next day.

nitrocode commented on runatlantis.io/docs/server-configuration.md:
```diff
@@ -456,6 +456,21 @@ Values are chosen in this order:
+  Please be mindful that settings like `--enable-diff-markdown-format` depend on logic defined in the templates. It is
+  possible to diverge from expected behavior, if care is not taken when overriding default templates.
+
+  Defaults to `~/.markdown_templates/`
```
What is `~` here ? is it the atlantis home directory ? If so, let's be explicit here.

pseudomorph replied:
Would explaining what `~/` is be sufficient (i.e. `'~/' represents the atlantis home dir`) or something like `/<atlantis_home_dir>/.markdown_templates/` ?

`--data-dir` uses the same notation in its description (just refers to `~/.atlantis`, but without explanation of what `~/` is.

Should I update `--data-dir` to reflect any explanation or expansion of `~/` I do here?

nitrocode replied:
This should be fine.

```suggestion
  Defaults to the atlantis home directory `/home/atlantis` in `$HOME/.markdown_templates/`
```

nitrocode commented on runatlantis.io/docs/server-configuration.md:
```diff
@@ -456,6 +456,21 @@ Values are chosen in this order:
+  Markdown template overrides may be specified either in individual files, or all together in a single file. All template
+  override files _must_ have the `.tmpl` extension, otherwise they will not be parsed.
+
+  Markdown templates which may have overrides can be found [here](../../server/events/templates)
```
This link fails to resolve when the documentation is setup. Are you trying to link to the code base ? If so, you should use an absolute link to the github page instead of the relative link.

nitrocode replied:
@pseudomorph

pseudomorph commented:
I noticed that some of the templates can be made more DRY. This is more of an optimization, and might not be an issue for now. It might take another day or two to get to that if it's desired.

Also, website_link_check is failing, and I'm not sure why.

ribejara-te commented:
Yeah I mostly kept templates as-is, only DRYing what I saw was easy, hoping we would get to further optimize them down the line. Feel free to do so. 🙂 

Btw thanks for pushing this down the finish line, really appreciated.

nitrocode commented on runatlantis.io/docs/server-configuration.md:
```diff
@@ -456,6 +456,21 @@ Values are chosen in this order:
+  Markdown template overrides may be specified either in individual files, or all together in a single file. All template
+  override files _must_ have the `.tmpl` extension, otherwise they will not be parsed.
+
+  Markdown templates which may have overrides can be found [here](https://github.com/runatlantis/atlantis/tree/master/server/events/templates)
```
This link fails for now but I think it will resolve once this gets merged. We can probably ignore this `branch` failure for now.

https://github.com/runatlantis/atlantis/tree/master/server/events/templates

@pseudomorph @ribejara-te

nitrocode reviewed (approved):
cc: @jamengual @chenrui333 please give this a second review

pseudomorph commented:
Thank you!

nitrocode commented:
Thank you for the contribution!

nitrocode commented:
Thank you for the contribution!

ribejara-te commented:
Thank you all for your help!
